Hash maps (#611)
fortran-lang/stdlib: doc/specs/stdlib_hash_maps.md

---
title: Hash maps
---

# The `stdlib_32_bit_key_data_wrapper`, `stdlib_chaining_hash_map` and `stdlib_open_hash_map` modules

(TOC)

## Overview of hash maps

The comparison of lexical entities or other objects for equality
can be computationally expensive.
This cost is often reduced by computing a near unique integer value,
termed a hash code, from the structure of the object using a procedure
termed a hash function.
Equality of hash codes is a necessary, but not sufficient, condition
for the original objects to be equal. 
As integer comparisons are very efficient, performing an initial
comparison of hash codes and then performing a detailed comparison
only if the hash codes are equal can improve performance.
The hash codes, in turn, can be mapped to a smaller set of integers,
that can be used as an index, termed a hash index, to a rank one
array, often termed a hash table or hash map.

> This document discusses the hash maps in the library.

Reviewer: Here I think the text initially gives too much of an introduction to hash codes, outside of the context of hash maps. It might be better to start out explaining what a hash map is (in sufficiently vague terms). 

Not sure what the wording should be. Perhaps: "A hash map is a data-structure that maps keys to values. Such data structures occur in many languages, for example the python `dict`, the C++ `map`, ...".

Author: Leet me see what I can come up with.

Author: I have revised that paragraph.

his code.

## The hash map modules

The Fortran Standard Library provides three modules for the
implementation of simple hash maps. These maps only accept hash
functions with a single argument, the key, and that yield a 32 bit
hash code, The modules will need to be modified to use hash functions
with a different API. There are three modules:
`stdlib_32_bit_key_data_wrapper`, `stdlib_chaining_hash_map` and
`stdlib_open_hash_map`, corresponding to the files:
`stdlib_32_bit_key_data_wrapper.f90`, `stdlib_chaining_hash_map.f90`,
and `stdlib_open_hash_map.f90`. The module
`stdlib_32_bit_key_data_wrapper` provides an interface to the 32 bit
hash functions of the Standard Library module,
`stdlib_32_bit_hash_functions`, and provides wrappers to some of the
hash functions so that they no longer need to be supplied seeds. The
module `stdlib_chaining_hash_map` defines a datatype,
`chaining_hash_map_type`, implementing a simple separate chaining hash
map noted more for its diagnostics than its performance. Finally the
module, `stdlib_open_hash_map` defines a datatype,
`open_hash_map_type`, implementing a simple open addressing hash

> map noted more for its diagnostics than its performance.

Reviewer: So both of the hash maps are `noted more for its diagnostics than its performance` ? That's OK, just want to double check that is the intention.

Author: I think the performance is not baad. Generating the information used in the diagnostics is, I suspect, a small amount of overhead, but it is an overhead other tables will lack.

largely defined in the separated submodules, `stdlib_chaining_hash_maps`
and `stdlib_open_hash_maps`, respectively.
In `chaining_hash_map_type` the colliding indices are handled by using
linked lists with their roots at the hash index.
In `open_hash_map_type`, the colliding indices are handled by searching
from the initial hash index in increasing
steps of one (modulo the hash map size) for an open map bin.

The maps share many attributes in common. The two types share a
common Application Programers Interface (API). The maps use powers of
two for their slot sizes, so that the function, `fibonacci_hash`, can
be used to map the hash codes to indices in the map. This is
expected to be more efficient than prime number mapping using a
modulo operation, and reduces the requirement that the hash
function need to do a good job randomizing its lower order bits.
This requires a good randomizing hash method for good performance.
Both adjust the map size to reduce collisions, based on 
the ratio of the number of hash map probes to the number of subroutine 
calls.
The maps make extensive use of pointers internally, but a private

> finalization subroutine avoids memory leaks.

Reviewer: Does this rely on the compiler to automatically call the finalization subroutine when a derived type goes out of scope? Last time I checked (albeit not recently) gfortran still didn't do this.

Author: Yes it relies on this to avoid memory leaks, though the largest amount of memory will be in arrays and not in the linked list used to avoid memory allocations. I don't think there are memory leaks with my most recent code compiled using the latest version of `gfortran`, though my testing was very indirect.

`key_type`, and `other_type`. The `key_type` is intended to be used
for the search keys of hash tables.  The `other_type` is intended to
store additional data associated with a key. Both types are
opaque. Their current representations are as follows

```fortran
    type :: key_type
        private
        integer(int8), allocatable :: value(:)
    end type key_type

    type :: other_type
        private
        integer(int8), allocatable :: value(:)
    end type other_type
```

The  module also defines seven procedures for those types: `copy_key`,
`copy_other`, `free_key`, `free_other`, `get`, `key_test`, and `set`
for use by the hash maps to manipulate or inquire of components of
those types.

### Table of `stdlib_32_bit_key_data_wrapper` procedures

The  `stdlib_32_bit_key_data_wrapper` module provides procedures in
several categories: procedures to manipulate data of the `key_type`;
procedures to manipulate data of the `other_type`, and 32 bit hash
functions for keys. The procedures in each category are listed below.

Procedures to manipulate `key_type` data:

* `copy_key( key_in, key_out )` - Copies the contents of the key,
  key_in, to the key, key_out.

> * `get( key, value )` - extracts the content of key into value.

Reviewer: Not immediately clear to me how this differs from `copy_key`

Reviewer: I see the details are explained below -- some minor edits here should make the difference clear.

Author: `copy_key` copies the internals of one key to the internals of another key. `get` copies the contents of a key or other data into an allocatable default character string or `int8` array. I have improved the documentation of this in `stdlib_hash_maps.md`.

Author: I added a few more details and pushed the result.

* `free_key( key )` - frees the memory in key.

* `set( key, value )` - sets the content of key to value.

* `key_test( key1, key2 )` - compares two keys for equality.

Procedures to manipulate `other_type` data:

* `copy_other( other_in, other_out )` - Copies the contents of the
  other data in, other_in, to the other data, other_out.

> * `get( other, value )` - extracts the content of other into value.

Reviewer: Not clear to me how this differs from `copy_other`

Author: See the above discussion for `copy_key` and `get`.

* `set( other, value )` - sets to content of other to value.

* `free_other( other )` - frees the memory in other.

Procedures to hash keys to 32 bit integers:

* `fnv_1_hasher( key )` - hashes a key using the FNV-1 algorithm.

* `fnv_1a_hasher( key )` - hashes a key using the FNV-1a algorithm.

* `seeded_nmhash32_hasher( key )` - hashes a key using the nmhash32
  algorithm.

* `seeded_nmhash32x_hasher( key )` - hashes a key using the nmhash32x
  algorithm.

* `seeded_water_hasher( key )` - hashes a key using the waterhash
  algorithm.

### Specifications of the `stdlib_32_bit_key_data_wrapper` procedures

#### `copy_key` - Returns a copy of the key

##### Status

Experimental

##### Description

Returns a copy of an input of type `key_type`

##### Syntax

`call [[stdlib_32_bit_key_data_wrapper:copy_key]]( key_in, key_out )`

##### Class

Subroutine.

##### Arguments

`key_in`: shall be a scalar expression of type `key_type`. It
is an `intent(in)` argument.

`key_out`: shall be a scalar variable of type `key_type`. It
is an `intent(out)` argument.

##### Example

```fortran
    program demo_copy_key
      use stdlib_32_bit_key_data_wrapper, only: &

>           copy_key, key_test, key_type

Reviewer: It looks like `key_test` is a test for equality? In that case consider if it would be better named `equal_keys`.

Author: > > The current use of `inmap` is about a factor if two more efficient than using `key` directly for modifications of the `other` data associated with the `key`. For initially entering an `entry`, testing for existence of the `key` in the `map` , and deleting an `entry` from the `map` there is not much difference in the efficiency between using `inmap` and using `key` directly.
> 
> If querying the key twice is a performance issue, could you return a `pointer` to other rather than its index `inmap`?

I would not return a pointer to internal data. It can cause a memory leak or other problems, and would require making more of the internal `targets` or `pointers`, making the structure more fragile. It is a performance issue for a lesser used aspect of hash maps, and I would rather make a localized performance degradation than use pointers.

Author: I don't know why this comment appeared here and not on the tail of the comments.

      use iso_fortran_env, only: int8
      implicit none
      integer(int8), allocatable :: value(:)
      type(key_type) :: key_in, key_out
      integer(int_8) :: i
      allocate( value(1:15) )
      do i=1, 15
          value(i) = i
      end do

>       call set( key, value )

Reviewer: Should `key` be `key_in` ?

Author: Yes. Fixed.

    end program demo_copy_key
```

#### `copy_other` - Returns a copy of the other data

##### Status

Experimental

##### Description

Returns a copy of an input of type `other_type`

##### Syntax

`call [[stdlib_32_bit_key_data_wrapper:copy_other]]( other_in, other_out )`

##### Class

Subroutine.

##### Arguments

`other_in`: shall be a scalar expression of type `other_type`. It
is an `intent(in)` argument.

`other_out`: shall be a scalar variable of type `other_type`. It
is an `intent(out)` argument.

##### Example

```fortran
    program demo_copy_other
      use stdlib_32_bit_key_data_wrapper, only: &
          copy_other, get, other_type, set
      use iso_fortran_env, only: int8
      implicit none
      integer(int8), allocatable :: value1(:), value2(:)
      type(other_type) :: other_in, other_out
      integer(int_8) :: i
      allocate( value1(1:15) )
      do i=1, 15
          value1(i) = i
      end do
      call set( other_in, value1 )
      call copy_other( other_in, other_out )
      call get( other_out, value2 )
      print *, "other_in == other_out = ", &
        all( value1 == value2 )
    end program demo_copy_other
```

> #### `FIBONACCI_HASH` - maps an integer to a smaller number of bits

Reviewer: Is this different to the version in `stdlib_hash_32bit`?

Author: It is  just giving access to the version in `stdlib_hash_32bi`.

Reviewer 2: If it is the case, then i suggest to to repeat the specs here. A link to the page of `Fibonacci_hash` should be enough,

Author: Do you mean "then i suggest not to repeat the specs here."?

Experimental

##### Description

Extracts the data from a `key_type` or an `other_type` and stores it
in the variable `value`..

##### Syntax

`call [[stdlib_32_bit_key_data_wrapper:get]]( key, value )`

or

`call [[stdlib_32_bit_key_data_wrapper:get]]( other, value )`

##### Class

Subroutine.

##### Argument

`key`: shall be a scalar expression of type `key_type`. It
is an `intent(in)` argument.

`other`: shall be a scalar expression of type `other_type`. It
is an `intent(in)` argument.

`value`: shall be an allocatable default character string variable, or
an allocatable vector variable of type integer and kind `INT8`. It is
an `intent(out)` argument.

##### Example

```fortran
    program demo_get
      use stdlib_32_bit_key_data_wrapper, only: &
          get, key_type, set
      use iso_fortran_env, only: int8
      implicit none
      integer(int8), allocatable :: value(:), result(:)
      type(key_type) :: key
      integer(int_8) :: i
      allocate( value(1:15) )
      do i=1, 15
        value(i) = i
      end do
      call set( key, value )
      call get( key, result )
      print *, `RESULT == VALUE = ', all( value  == result )
    end program demo_get
```

> #### `HASHER_FUN`- serves aa function prototype.

Reviewer: typo

Author: Fixed several typos in the documentation for `hasher_fun`.

##### Description

Serves as a prototype for hashing functions with a single, `key`,
argument returning an `INT322` hash value.

##### Syntax

`type([[stdlib_32_bit_key_data_wrapper:hasher_fun]]), pointer :: fun_pointer`

##### Class

Pure function prototype

##### Argument

`key`: Shall be a rank one array expression of type `INTEGER(INT8)`.
It is an `intent(in)` argument.

##### Result character

The result is a scalar integer of kind `INT32`.

##### Result value

The result is a hash code.

##### Note

`HASHER_FUN` is a prototype for defining dummy arguments and function
pointers intended for use 

##### Example

```fortran
    program demo_hasher_fun
      use stdlib_32_bit_key_data_wrapper, only: &
          fnv_1a_hasher, hasher_fun, set
      use iso_fortran_env, only: int8, int32 
      implicit none
      type(hasher_fun), pointer :: hasher_pointer
      integer(int8), allocatable :: array1(:)
      integer(int32) :: hash
      type(key_type) :: key
      hasher_pointer => fnv_1a_hasher
      array1 = [ 5_int8, 4_int8, 3_int8, 1_int8, 10_int8, 4_int8 ]
      call set( key, array1 )
      hash = hassher_pointer(key)
      print *, hash
    end program demo_hasher_fun
```

> #### `key_test` - Compares two keys for equality

Reviewer: As above, consider renaming to `equal_keys` or `keys_equal` or similar.

## Overview of hash maps

A hash map (hash table) is a data structure that maps *keys* to
*values*. It uses a hash function to compute a hash code from the *key*
that serves as an index into a linear array of *slots* (buckets) from
which the desired *value* can be extracted.
Each key ideally maps to a unique slot, but most hash functions are
imperfect and can map multiple keys to the same *slot* resulting in
collisions. Hash maps differ in how they deal with such collisions.
This document discusses the hash maps in the Fortran Standard Library.

## Licensing

The Fortran Standard Library is distributed under the MIT License.
However components of the library should be evaluated as to whether
they are compatible with the MTI License.
The current hash maps were inspired by an
[implementation](http://chasewoerner.org/src/hasht/) of David
Chase. While the code has been greatly modified from his
implementation, he has give permission for the unrestricted use of
his code.

## The hash map modules

The Fortran Standard Library provides three modules for the
implementation of simple hash maps. These maps only accept hash
functions with a single argument, the key, and that yield a 32 bit

> hash code, The modules will need to be modified to use hash functions

Author: I have greatly revised that paragraph.

with a different API. There are three modules:
`stdlib_hashmap_wrappers`, `stdlib_hashmap_chaining` and
`stdlib_hashmap_open`, corresponding to the files:
`stdlib_hashmap_wrappers.f90`, `stdlib_hashmap_chaining.f90`,
and `stdlib_hashmap_open.f90`. The module
`stdlib_hashmap_wrappers` provides an interface to the 32 bit
hash functions of the Standard Library module,
`stdlib_hash_32bit`, and provides wrappers to some of the
hash functions so that they no longer need to be supplied seeds. The
module `stdlib_hashmap_chaining` defines a datatype,

> `chaining_hash_map_type`, implementing a simple separate chaining hash

Author: I changed all  mentions of `hash_map` to `hash map`.

map noted more for its diagnostics than its performance. Finally the

> module, `stdlib_hashmap_open` defines a datatype,

Author: Fixed

`open_hash_map_type`, implementing a simple open addressing hash
map noted more for its diagnostics than its performance.

These maps use separate chaining with linked lists and linear open
addressing, respectively, to deal with hash index collisions, and are
largely defined in the separated submodules, `stdlib_hashmap_chainings`
and `stdlib_hashmap_opens`, respectively.

> In `chaining_hash_map_type` the colliding indices are handled by using

Author: Fixed see above.

linked lists with their roots at the hash index.

> In `open_hash_map_type`, the colliding indices are handled by searching

Author: See above.

from the initial hash index in increasing
steps of one (modulo the hash map size) for an open map bin.

The maps share many attributes in common. The two types share a
common Application Programers Interface (API). The maps use powers of
two for their slot sizes, so that the function, `fibonacci_hash`, can
be used to map the hash codes to indices in the map. This is
expected to be more efficient than prime number mapping using a
modulo operation, and reduces the requirement that the hash
function need to do a good job randomizing its lower order bits.
This requires a good randomizing hash method for good performance.
Both adjust the map size to reduce collisions, based on 
the ratio of the number of hash map probes to the number of subroutine 
calls.
The maps make extensive use of pointers internally, but a private
finalization subroutine avoids memory leaks.
The maps can take entry keys of type `key_type`.
Both maps allow the addition and lookup of entries, and the inclusion
of data in addition to the entry key.

> The `chaining_hash_map_type` also allows the selective removal of

Author: See above

entries.

## The `stdlib_hashmap_wrappers` module

The `stdlib_hashmap_wrappers` module provides data types to
represent keys and associated data stored in a module, but is also, a
wrapper for the `stdlib_hash_32bit` module. It allows
direct access to the `stdlib_hash_32bit` procedures:

> `fibonacci_hash`, `fnv_1_hasher`, `fnv_1a_hasher`; and provides

Reviewer 2: ```suggestion
`fibonacci_hasher`, `fnv_1_hasher`, `fnv_1a_hasher`; and provides
```

Author: No it should be `fibonacci_hash`

wrapper functions, `seeded_nmhash32_hasher`,
`seeded_nmhash32x_hasher`, and `seeded_water_hasher` to the hashing
functions: `nmhash32`, `nmhash32x`, and `water_hash`, respectively. It
defines an interface, `hasher_fun`, compatible with the hash functions
that take a `non-scalar key`. It defines one integer constant used
as a kind value,`int_hash`. It also defines two types, `key_type` and
`other_type`, and associated procedures, for storing and manipulating
keys and their associated data.

> ### The `stdlib_hashmap_wrappers` constant, `INT_HASH`

Reviewer 2: ```suggestion
### The `stdlib_hashmap_wrappers` constant, `int_hash`
```

Author: I have tried to change all upper case names to lower case. Let me know if I missed any.

The constant `INT_HASH` is used to define the integer kind value for
the returned hash codes and variables used to access them. It

> currently has the value, `INT32`.

Reviewer 2: ```suggestion
The constant `int_hash` is used to define the integer kind value for
the returned hash codes and variables used to access them. It
currently has the value, `int32`.
```

Author: See above,

* `equal_keys( key1, key2 )` - compares two keys for equality. 

* `get( key, value )` - extracts the contents of key into value, an
  `int8` array or character string.

* `free_key( key )` - frees the memory in key.

* `set( key, value )` - sets the content of key to value.

Procedures to manipulate `other_type` data:

* `copy_other( other_in, other_out )` - Copies the contents of the
  other data, `other_in`, to the contents of the other data,
  `other_out`.

* `get( other, value )` - extracts the contents of other into value, an 
  `int8` array or character string. 

* `set( other, value )` - sets to content of other to value.

* `free_other( other )` - frees the memory in other.

Procedures to hash keys to 32 bit integers:

* `fnv_1_hasher( key )` - hashes a key using the FNV-1 algorithm.

* `fnv_1a_hasher( key )` - hashes a key using the FNV-1a algorithm.

* `seeded_nmhash32_hasher( key )` - hashes a key using the nmhash32
  algorithm.

* `seeded_nmhash32x_hasher( key )` - hashes a key using the nmhash32x
  algorithm.

* `seeded_water_hasher( key )` - hashes a key using the waterhash
  algorithm.

### Specifications of the `stdlib_hashmap_wrappers` procedures

> #### `copy_key` - Returns a copy of the key

Reviewer 2: could it overload "="?

Author: Yes it could. It would require exchanging argument locations and changing all references to `copy_key` in the code. The same could be done for `copy_other`.

##### Example

```fortran
    program demo_copy_other
      use stdlib_hashmap_wrappers, only: &
          copy_other, get, other_type, set
      use iso_fortran_env, only: int8
      implicit none
      integer(int8), allocatable :: value1(:), value2(:)
      type(other_type) :: other_in, other_out
      integer(int_8) :: i
      allocate( value1(1:15) )
      do i=1, 15
          value1(i) = i
      end do
      call set( other_in, value1 )
      call copy_other( other_in, other_out )
      call get( other_out, value2 )
      print *, "other_in == other_out = ", &
        all( value1 == value2 )
    end program demo_copy_other
```

#### `equal_keys` - Compares two keys for equality

##### Status

Experimental

##### Description

Returns `.true.` if two keys are equal, and false otherwise.

##### Syntax

`test = [[stdlib_hashmap_wrappers:equal_keys]]( key1, key2 )`

##### Class

Pure function.

##### Arguments

`key1`: shall be a scalar expression of type `key_type`. It
is an `intent(in)` argument.

`key2`: shall be a scalar expression of type `key_type`. It
is an `intent(in)` argument.

##### Result character

The result is a value of type default `LOGICAL`.

##### Result value

> The result is `.TRUE.` if the keys are equal, otherwise `.FALSS`.

Reviewer 2: ```suggestion
The result is `.true.` if the keys are equal, otherwise `.false.`.
```

Author: see above

##### Example

```fortran
    program demo_free_other
      use stdlib_hashmap_wrappers, only: &
          copy_other, free_other, other_type, set
      use iso_fortran_env, only: int8
      implicit none
      integer(int8), allocatable :: value(:)
      type(key_type) :: other_in, other_out
      integer(int_8) :: i
      allocate( value(1:15) )
      do i=1, 15
          value(i) = i
      end do
      call set( other_in, value )
      call copy_other( other_in, other_out )
      call free_other( other_out )
    end program demo_free_other
```

#### `get` - extracts the data from a derived type

##### Status

Experimental

##### Description

Extracts the data from a `key_type` or an `other_type` and stores it
in the variable `value`..

##### Syntax

`call [[stdlib_hashmap_wrappers:get]]( key, value )`

or

`call [[stdlib_hashmap_wrappers:get]]( other, value )`

##### Class

Subroutine.

##### Argument

`key`: shall be a scalar expression of type `key_type`. It
is an `intent(in)` argument.

`other`: shall be a scalar expression of type `other_type`. It
is an `intent(in)` argument.

`value`: shall be an allocatable default character string variable, or

> an allocatable vector variable of type integer and kind `INT8`. It is

Reviewer 2: ```suggestion
an allocatable vector variable of type integer and kind `int8`. It is
```

Author: Done.

an `intent(out)` argument.

##### Example

```fortran
    program demo_get
      use stdlib_hashmap_wrappers, only: &
          get, key_type, set
      use iso_fortran_env, only: int8
      implicit none
      integer(int8), allocatable :: value(:), result(:)
      type(key_type) :: key
      integer(int_8) :: i
      allocate( value(1:15) )
      do i=1, 15
        value(i) = i
      end do
      call set( key, value )
      call get( key, result )
      print *, `RESULT == VALUE = ', all( value  == result )
    end program demo_get
```

#### `HASHER_FUN`- serves aa a function prototype.

##### Status

Experimental

##### Description

Serves as a prototype for hashing functions with a single, `key`,

> argument returning an `INT32` hash value.

Reviewer 2: ```suggestion
argument returning an `int32` hash value.
```

Author: Done.

##### Syntax

`type([[stdlib_hashmap_wrappers:hasher_fun]]), pointer :: fun_pointer`

##### Class

Pure function prototype

##### Argument

> `key`: Shall be a rank one array expression of type `INTEGER(INT8)`.

Reviewer 2: ```suggestion
`key`: Shall be a rank one array expression of type `integer(int8)`.
```

Author: > Done.

It is an `intent(in)` argument.

##### Result character

> The result is a scalar integer of kind `INT32`.

Reviewer 2: ```suggestion
The result is a scalar integer of kind `int32`.
```

Author: Done

##### Result value

The result is a hash code.

##### Note

`HASHER_FUN` is a prototype for defining dummy arguments and function
pointers intended for use as a hash function for the hash maps.

##### Example

```fortran
    program demo_hasher_fun
      use stdlib_hashmap_wrappers, only: &
          fnv_1a_hasher, hasher_fun, set
      use iso_fortran_env, only: int8, int32 
      implicit none
      type(hasher_fun), pointer :: hasher_pointer
      integer(int8), allocatable :: array1(:)
      integer(int32) :: hash
      type(key_type) :: key
      hasher_pointer => fnv_1a_hasher
      array1 = [ 5_int8, 4_int8, 3_int8, 1_int8, 10_int8, 4_int8 ]
      call set( key, array1 )
      hash = hassher_pointer(key)
      print *, hash
    end program demo_hasher_fun
```

#### `SEEDED_NMHASH32_HASHER`- calculates a hash code from a key

##### Status

Experimental

##### Description

Calculates a 32 bit hash code from an input of type `key_type`.

##### Syntax

`code = [[stdlib_hashmap_wrappers:seeded_nmhash32_hasher]]( key )`

##### Class

Pure function

##### Argument

`key`: Shall be a scalar expression of type `key_type`.
It is an `intent(in)` argument.

##### Result character

> The result is a scalar integer of kind `INT32`.

Reviewer 2: ```suggestion
The result is a scalar integer of kind `int32`.
```

Author: Done

##### Result value

The result is a hash code created using the `nmhash32` algorithm.

##### Note

`SEEDED_NMHASH32_HASHER` is a wrapper to the `NMHASH32_HASH` of the
module `stdlib_hash_32bit`, which supplies a fixed seed
to the wrapped function. `NMHASH32` is an implementation of the
`nmhash32` hash code of James Z. M. Gao.
This code has good, but not great, performance on long keys, poorer
performance on short keys.
As a result it should give fair performance for typical hash map
applications.
This code passes the SMHasher tests.
As a result it should give good performance for typical hash map

> applications.

Reviewer 2: Repeated sentence

```suggestion
```

Author: Done.

      hash = seeded_nmhash32_hasher (key)
      print *, hash
    end program demo_seeded_nmhash32_hasher
```

#### `SEEDED_NMHASH32X_HASHER`- calculates a hash code from a key

##### Status

Experimental

##### Description

Calculates a 32 bit hash code from an input of type `key_type`.

##### Syntax

`code = [[stdlib_hashmap_wrappers:seeded_nmhash32x_hasher]]( key )`

##### Class

Pure function

##### Argument

`key`: Shall be a scalar expression of type `key_type`.
It is an `intent(in)` argument.

##### Result character

The result is a scalar integer of kind `INT32`.

##### Result value

The result is a hash code created using the `nmhash32x` algorithm.

##### Note

`SEEDED_NMHASH32X_HASHER` is a wrapper to the `NMHASH32X_HASH` of the
module `stdlib_hash_32bit`, which supplies a fixed seed
to the wrapped function. `NMHASH32X` is an implementation of the
`nmhash32x` hash code of James Z. M. Gao.
This code has good, but not great, performance on long keys, poorer
performance on short keys.
As a result it should give fair performance for typical hash map
applications.
This code passes the SMHasher tests.
As a result it should give good performance for typical hash map

> applications.

Reviewer 2: ```suggestion
```

Author: Done

      hash = seeded_nmhash32x_hasher (key)
      print *, hash
    end program demo_seeded_nmhash32x_hasher
```

#### `SEEDED_WATER_HASHER`- calculates a hash code from a key

##### Status

Experimental

##### Description

Calculates a 32 bit hash code from an input of type `key_type`.

##### Syntax

`code = [[stdlib_hashmap_wrappers:seeded_water_hasher]]( key )`

##### Class

Pure function

##### Argument

`key`: Shall be a scalar expression of type `key_type`.
It is an `intent(in)` argument.

##### Result character

The result is a scalar integer of kind `INT32`.

##### Result value

The result is a hash code created using the `waterhash` algorithm.

##### Note

`SEEDED_WATER_HASHER` is a wrapper to the `WATER_HASH` of the
module `stdlib_hash_32bit`, which supplies a fixed seed
to the wrapped function. `WATER_HASH` is an implementation of the
`waterhash` hash code of Tommy Ettinger.
This code has excellent performance on long keys, and good performance
on short keys.
As a result it should give reasonable performance for typical hash
table applications.
This code passes the SMHasher tests.
As a result it should give good performance for typical hash map

> applications.

Reviewer 2: ```suggestion
```

Author: Done.

Subroutine.

##### Argument

`key`: shall be a scalar variable of type `key_type`. It
is an `intent(out)` argument.

`other`: shall be a scalar variable of type `other_type`. It
is an `intent(out)` argument.

`value`: shall be a default character string expression, or a
vector expression of type integer and kind `INT8`. It is an
`intent(in)` argument.

##### Example

```fortran
    program demo_set
      use stdlib_hashmap_wrappers, only: &
          get, key_type, set
      use iso_fortran_env, only: int8
      implicit none
      integer(int8), allocatable :: value(:), result(:)
      type(key_type) :: key
      integer(int_8) :: i
      allocate( value(1:15) )
      do i=1, 15
        value(i) = i
      end do
      call set( key, value )
      call get( key, result )
      print *, `RESULT == VALUE = ', all( value  == result )
    end program demo_set
```

## The `stdlib_hashmap_chaining` module

The `stdlib_hashmap_chaining` module provides access to all the
public entities in the `stdlib_hashmap_wrappers` module. It
also defines a public data type and associated procedures and
constants that implement a simple hash map using
separate chaining hashing. The derived type is

> `chaining_hash_map_type`. It provides 

Reviewer 2: ```suggestion
`chaining_hashmap_type`. It provides 
```

Author: Done.

procedures to manipulate the structure of the hash map:
`init`, `map_entry`, `rehash`, `remove_entry`, and
`set_other_data`. It provides procedures to inquire about entries in

> the hash map: `get_other_data`, `in_map`, `unmap`.and `valid_index`.

Reviewer 2: ```suggestion
the hash map: `get_other_data`, `in_map`, `unmap` and `valid_index`.
```

Author: Done.

Finally it provides procedures to inquire about the overall
structure and performance of the table:`calls`,  `entries`,
`get_other_data`, `loading`, `slots`, and `total_depth`. The module
also defines a number of public constants: `inmap_probe_factor`,
`map_probe_factor`, `default_bits`,
`max_bits`, `int_calls`, `int_depth`, `int_index`,
`int_probes`, `success`, `alloc_fault`, and `array_size_error`.

### The `stdlib_hashmap_chaining` module's public constants

The module defines several categories of public constants. Some are
used to parameterize the empirical slot expansion code. Others

> parameterize the slots table size, Some are used to define

Reviewer 2: ```suggestion
parameterize the slots table size. Some are used to define
```

Author: Done.

parameterize the slot expansion code used to determine when in a
`inchain_map_call` the number 
of slots need to be increased to decrease the lengths of the linked
lists. The constant `inmap_probe_factor` is used to determine when
the ratio of the number of map probes to map calls is too large and
the slots need expansion. The constant `map_probe_factor` is used to
determine when inserting a new entry the ratio of the number of map
probes to map calls is too large and the slots need expansion.

The constants `default_bits`, and
`max_bits` are used to parameterize the table's slots size. The
`default_bits` constant defines the default initial number of slots
with a current value of 6 resulting in an initial `2**6 == 64`
slots. This may optionally be overridden on hash map creation. The
`max_bits` parameter sets the maximum table size as `2**max_bits` with
a default value for `max_bits` of 30. The table will not work for a
slots size greater than `2**30`.

The constants `int_calls`, `int_depth`, `int_index`, and `int_probes`
are used to define integer kind values for various contexts. The
number of calls are reported and stored in entities of kind

> `int_calls`. Currently `int_calls` has the value of `INT64`. The

Reviewer 2: ```suggestion
`int_calls`. Currently `int_calls` has the value of `int64`. The
```

Author: Done.

total depth, the number of inquiries needed to access all elements
of the table, is reported and stored in entities of kind

> `int_depth`. Currently `int_depth` has the value of `INT64`. The

Reviewer 2: ```suggestion
`int_depth`. Currently `int_depth` has the value of `int64`. The
```

Author: Done.

number of entries in the table, is reported and stored in entities of

> kind `int_index`. Currently `int_index` has the value of `INT32`.

Reviewer 2: ```suggestion
kind `int_index`. Currently `int_index` has the value of `int32`.
```

Author: Done.

The number of probes, hash map enquiries, are reported and stored in
entities of kind `int_probes`. Currently `int_probes` has the value of
`INT64`.

Finally the error codes `success`, `alloc_fault`, and
`array_size_error` are used to report the error status of certain

> procedure calla. The `succes` code indicates that no problems were

Reviewer 2: ```suggestion
procedure calls. The `succes` code indicates that no problems were
```

Author: Done.

found. The `alloc_fault` code indicates that a memory allocation
failed. Finally the `array_size_error` indicates that on table
creation `slots_bits` is less than `default_bits` or
greater than `max_bits`.

### The `stdlib_hashmap_chaining` module's derived types

The `stdlib_hashmap_chaining` module defines several derived

> types. The only public type is the `chaining_hash_map_type`. There are

Reviewer 2: ```suggestion
types. The only public type is the `chaining_hashmap_type`. There are
```

Author: Done.

three other private derived types used in the implementation of the
public type: `chaining_map_entry_type`, `chaining_map_entry_ptr`, and
`chaining_map_entry_pool`. Each of these is described below.

#### The `chaining_map_entry_type` derived type

Entities of the type `chaining_map_entry_type` are used to define
a linked list structure that stores the
key, its other data, the hash of the key, and the resulting index into
the inverse table. The type's definition is below:

```fortran
    type :: chaining_map_entry_type  ! Chaining hash map entry type
        private
        integer(int_hash)   :: hash_val ! Full hash value
        type(key_type)      :: key ! The entry's key
        type(other_type)    :: other ! Other entry data
        integer(int_index)  :: index ! Index into inverse table
        type(chaining_map_entry_type), pointer :: &
            next => null() ! Next bucket
    end type chaining_map_entry_type
```

> Currently the `INT_HASH` and `INT_INDEX` have the value of `INT32`.

Reviewer 2: ```suggestion

Currently the `int_hash` and `int_index` have the value of `int32`.
```

Author: Done.

#### The `chaining_map_entry_ptr` derived type

> The type `chaining_map_entry_ptr` are used to define the elements of

Reviewer 2: ```suggestion
The type `chaining_map_entry_ptr` is used to define the elements of
```

Author: Fixed at both locations.

the hash map that are either empty or link to the linked lists
containing the elements of the table. The type's definition is below:

```fortran
    type chaining_map_entry_ptr ! Wrapper for a pointer to a chaining
                                ! map entry type object
        type(chaining_map_entry_type), pointer :: target => null()
    end type chaining_map_entry_ptr
```

#### The `chaining_map_entry_pool` derived type

The type `chaining_map_entry_pool` is used to implement a pool of
allocated `chaining_map_entry_type` elements to save on allocation
costs. The type's definition is below: 

```fortran
    type :: chaining_map_entry_pool

>     ! Type inplementing a pool of allocated

Reviewer 2: ```suggestion
    ! Type implementing a pool of allocated
```

Author: Fixed both locations

  slots in a hash map.

* `map_probes( map )` - the total number of table probes on a hash
  map.

* `slots( map )` - Returns the number of allocated slots in a hash
  map.

* `total_depth( map )` - Returns the total number of one's based
offsets of slot entries from their slot index

### Specifications of the `stdlib_hashmap_chaining` procedures

#### `calls` - Returns the number of calls on a hash map

##### Status

Experimental

##### Description

Returns the number of procedure calls on a hash map.

##### Syntax

`value = [[stdlib_hashmap_chaining:calls]]( map )`

##### Class

Pure function

##### Argument

`map` - shall be an expression of type `chaining_hash_map_type`.
It is an `intent(in)` argument.

##### Result character

The result will be an integer of kind `INT_CALLS`.

##### Result value

The result will be the number of procedure calls on the hash map.

##### Example

```fortran
    program demo_calls
      use stdlib_hashmap_chaining, only: &
         chaining_hash_map_type, calls, init, int_calls, &
         fnv_1_hasher
      implicit none
      type(chaining_hash_map_type) :: map
      type(int_calls) :: initial_calls
      call init( map, fnv_1_hasher )

>       initisl_calls = calls (map)

Reviewer 2: ```suggestion
      initial_calls = calls (map)
```

Author: Fixed both locations.

      print *, "INITIAL_CALLS =  ", initial_calls
    end program demo_calls
```

#### `entries` - Returns the number of entries in a hash map

##### Status

Experimental

##### Description

Returns the number of entries in a hash map.

##### Syntax

`value = [[stdlib_hashmap_chaining:entries]]( map )`

##### Class

Pure function

##### Argument

`map` - shall be an expression of type `chaining_hash_map_type`.
It is an `intent(in)` argument.

##### Result character

> The result will be an integer of kind `INT_INDEX`.

Reviewer 2: ```suggestion
The result will be an integer of kind `int_index`.
```

Author: done.

##### Result value

The result will be the number of entries in the hash map.

##### Example

```fortran
    program demo_entries
      use stdlib_hashmap_chaining, only: &
         chaining_hash_map_type, entries, init, int_index, &
         fnv_1_hasher
      implicit none
      type(chaining_hash_map_type) :: map
      type(int_index) :: initial_entries
      call init( map, fnv_1_hasher )

>       initisl_entries = entries (map)

Reviewer 2: ```suggestion
      initial_entries = entries (map)
```

Author: Done.

            int8
        use stdlib_hashmap_chaining, only: &
            chaining_hash_map_type, fnv_1_hasher, in_map, &
            int_index, key_type, map_entry, other_type, set
        integer(int_index)           :: inmap
        type(key_type)               :: key
        type(other_type)             :: other
        type(chaining_hash_map_type) :: map
        call init( map, fnv_1_hasher )
        call set( key, [ 0_int8, 1_int8, 2_int8, 3_int8, 4_int8 ] )
        call set( other, [ 4_int8, 3_int8, 2_int8, 1_int8 ] )
        call map_entry( map, inmap, key, other )
        if ( inmap /= 0 ) then
            call in_map( map, inmap, key
            if ( inmap \= 0 ) then
			   print *, 'INMAP = ', inmap
            else
                stop 'Invalid inmap from in_map call'
        else
            stop 'Invalid inmap from map_entry call'
        end if
    end program demo_in_map
```

#### init - initializes a hash map

##### Status

Experimental

##### Description

Initializes a `chaining_hash_map_type` object.

##### Syntax

`call [[stdlib_hashmap_chaining:init]](  map, hasher [, slots_bits, status ] ] )`

####@# Class

Subroutine

##### Arguments

> `map`): shall be a scalar variable of type

Reviewer 2: ```suggestion
`map`: shall be a scalar variable of type
```

Author: Done.

`hasher`: shall be a procedure with interface `hash_fun`.
  It is an `intent(in)` argument. It is the procedure to be used to
  generate the hashes for the table from the keys of the entries.

`slots_bits` (optional): shall be a scalar default integer 
  expression. It is an `intent(in)` argument. The initial number of
  slots in the table will be `2**slots_bits`.

* `slots_bits` shall be a positive default integer less than
  `max_bits`, otherwise processing stops with an informative
  error code.

* If `slots_bits` is absent then the effective value for `slots_bits`
  is `default_slots_bits`.

`status` (optional): shall be a scalar integer variable of kind
`int32`. It is an `intent(out)` argument. On return if present it
shall have an error code value.

* If map was successfully initialized then `status` has the value
`success`.

* If allocation of memory for the `map` arrays fails then `status`
has the value `alloc_fault`.

* If `slot_bits < 6` or `slots_bits > max_bits` then `status`
  has the value of `array_size_error`.

* If `status` is absent, but `status` would have a value other than
`success`, then processing stops with an informative stop code.

##### Example

>     program demo_init

Reviewer 2: ````suggestion
```fortran
    program demo_init
````

Author: Done.

        use stdlib_hash_tables, only: &
            chaining_map_type, fnv_1_hasher &
            init
        type(fnv_1a_type)       :: fnv_1
        type(chaining_map_type) :: map
        call init( map,           &
                   fnv_1a,        &

>                    slots_bits=10 )

Reviewer 2: ````suggestion
                   slots_bits=10 )
```
````

Author: Done.

    end program demo_init

#### `loading` - Returns the ratio of entries to slots

##### Status

Experimental

##### Description

Returns the ratio of the number of entries relative to the number of
slots in a hash map.

##### Syntax

`value = [[stdlib_hashmap_chaining:loading]]( map )`

##### Class

Pure function

##### Argument

`map` - shall be an expression of type `chaining_hash_map_type`.
It is an `intent(in)` argument.

##### Result character

The result will be a default real.

##### Result value

The result will be the ratio of the number of entries relative to the

> number of slots in the hash map.?

Reviewer 2: ```suggestion
number of slots in the hash map.
```

Author: Done at both locations.

##### Syntax

`Result = [[stdlib_hashmap_chaining:total_depth]]( map )`

##### Class

Pure function

##### Argument

`map`: shall be a scalar expression of type
`chaining_hash_map_type`. It is an `intent(in)` argument. It is the
hash map of interest.

##### Result character

The result is a scalar integer of kind `int_depth`.

##### Result value

The result is the total number of one's based offsets of slot entries
from their slot index the map.

##### Example

```fortran
    program demo_probes
      use stdlib_hashmap_chaining, only: &
         chaining_hash_map_type, init, int_index, &
         fnv_1_hasher, total_depth
      implicit none
      type(chaining_hash_map_type) :: map
      integer(int_depth) :: initial_depth
      call init( map, fnv_1_hasher )
      initial_depth = total_depth (map)
      print *, "Initial total depth =  ", initial_depth
    end program demo_probes
```

#### `unmap` - returns a copy of the key

##### Status

Experimental

##### Description

Returns a copy of the key associated with an index to the 
inverse table.

##### Syntax

`call [[stdlib_hashmap_chaining:unmap]]( map, inmap, key )`

##### Class

Subroutine

##### Arguments

> `map`: shall be a scalar expression of type `chaining_hash_map_type.

Reviewer 2: ```suggestion
`map`: shall be a scalar expression of type `chaining_hash_map_type`.
```

Author: Fixed the two instances.

  map.

* `relative_loading` - the ratio of the map's loading to its
  `load_factor`. 

* `slots( map )` - Returns the number of allocated slots in a hash
  map.

* `total_depth( map )` - Returns the total number of one's based
offsets of slot entries from their slot index

### Specifications of the `stdlib_hashmap_open` procedures

#### `calls` - Returns the number of calls on a hash map

##### Status

Experimental

##### Description

Returns the number of procedure calls on a hash map.

##### Syntax

`value = [[stdlib_hashmap_open:calls]]( map )`

##### Class

Pure function

##### Argument

`map` - shall be an expression of type `open_hash_map_type`.
It is an `intent(in)` argument.

##### Result character

The result will be an integer of kind `INT_CALLS`.

##### Result value

The result will be the number of procedure calls on the hash map.

##### Example

```fortran
    program demo_calls
      use stdlib_hashmap_open, only: &
         open_hash_map_type, calls, init, int_calls, &
         fnv_1_hasher
      implicit none
      type(open_hash_map_type) :: map
      type(int_calls) :: initial_calls
      call init( map, fnv_1_hasher )

>       initisl_calls = calls (map)

Reviewer 2: ```suggestion
      initial_calls = calls (map)
```

Author: Fixed the two instances.

      print *, "INITIAL_CALLS =  ", initial_calls
    end program demo_calls
```

#### `entries` - Returns the number of entries in a hash map

##### Status

Experimental

##### Description

Returns the number of entries in a hash map.

##### Syntax

`value = [[stdlib_hashmap_open:entries]]( map )`

##### Class

Pure function

##### Argument

`map` - shall be an expression of type `open_hash_map_type`.
It is an `intent(in)` argument.

##### Result character

The result will be an integer of kind `INT_INDEX`.

##### Result value

The result will be the number of entries in the hash map.

##### Example

```fortran
    program demo_entries
      use stdlib_hashmap_open, only: &
         open_hash_map_type, entries, init, int_index, &
         fnv_1_hasher
      implicit none
      type(open_hash_map_type) :: map
      type(int_index) :: initial_entries
      call init( map, fnv_1_hasher )

>       initisl_entries = entries (map)

Reviewer 2: ```suggestion
      initial_entries = entries (map)
```

Author: Done.

            int8
        use stdlib_hashmap_open, only: &
            open_hash_map_type, fnv_1_hasher, in_map, &
            int_index, key_type, map_entry, other_type, set
        integer(int_index)           :: inmap
        type(key_type)               :: key
        type(other_type)             :: other
        type(open_hash_map_type) :: map
        call init( map, fnv_1_hasher )
        call set( key, [ 0_int8, 1_int8, 2_int8, 3_int8, 4_int8 ] )
        call set( other, [ 4_int8, 3_int8, 2_int8, 1_int8 ] )
        call map_entry( map, inmap, key, other )
        if ( inmap /= 0 ) then
            call in_map( map, inmap, key
            if ( inmap \= 0 ) then
			   print *, 'INMAP = ', inmap
            else
                stop 'Invalid inmap from in_map call'
        else
            stop 'Invalid inmap from map_entry call'
        end if
    end program demo_in_map
```

#### init - initializes a hash map

##### Status

Experimental

##### Description

Initializes a `open_hash_map_type` object.

##### Syntax

`call [[stdlib_hashmap_open:init]](  map, hasher[, slots_bits, status ] ]`

####@# Class

Subroutine

##### Arguments

> `map`): shall be a scalar variable of type

Reviewer 2: ```suggestion
`map`: shall be a scalar variable of type
```

Author: Fixed the two instances.

`open_hash_map_type`. It is an `intent(in)` argument. It is the
hash map of interest.

##### Result character

The result is a scalar integer of kind `int_probes`.

##### Result value

The result is the number of probes of `map`.

##### Example

```fortran
    program demo_probes
      use stdlib_hashmap_open, only: &
         open_hash_map_type, init, int_index, &
         fnv_1_hasher, probes
      implicit none
      type(open_hash_map_type) :: map
      real :: ratio
      call init( map, fnv_1_hasher )
      ratio = probes (map)
      print *, "Initial probes =  ", ratio
    end program demo_probes
```

#### rehash - changes the hashing function

##### Status

Experimental

##### Description

Changes the hashing function for the table entries to that of `hasher`.

##### Syntax

`call [[stdlib_hashmap_open:rehash]]( map, hasher )`

##### Class

Subroutine

##### Arguments

`map` : shall be a scalar variable of type `open_hash_map_type`.
It is an `intent(inout)` argument. It is the hash map whose hashing 
method is to be changed.

`hasher`: shall be a function of interface `hasher_fun`.
It is the hash method to be used by `map`.

##### Example

>     program demo_rehash

Author: It also needs the closing ```. Done

        call set( key, [ 5_int8, 7_int8, 4_int8, 13_int8 ] )
        call set( other, [ 1_int8, 5_int8, 3_int8, 15_int8 ] )
        call map_entry( map, inmap, key, other )
        call rehash( map, fnv_1a_hasher )
    end program demo_rehash

#### `relative_loading` - Returns the ratio of `loading` to `load_factor`

##### Status

Experimental

##### Description

Returns the ratio of the loadings relative to the open hash map's
`load_factor`. 

##### Syntax

`value = [[stdlib_hashmap_open:relative_loading]]( map )`

##### Class

Pure function

##### Argument

`map` - shall be an expression of type `open_hash_map_type`.
It is an `intent(in)` argument.

##### Result character

The result will be a default real.

##### Result value

The result will be the ratio of the number of entries relative to the
number of slots in the hash map relative to the `load_factor`.

##### Example

```fortran
    program demo_relative_loading
      use stdlib_hashmap_open, only: &
         open_hash_map_type, init, int_index, &
         fnv_1_hasher, loading
      implicit none
      type(open_hash_map_type) :: map
      real :: ratio
      call init( map, fnv_1_hasher )

>       ratio = relative loading (map)

Reviewer 2: ```suggestion
      ratio = loading (map)
```

Author: Should be `ratio = relative_loading`.

##### Status

Experimental

##### Description

Replaces the other data for the entry at index `inmap` in the 
inverse table.

##### Syntax

`call [[stdlib_hashmap_open:set_other_data]]( map, inmap, other )`

##### Class

Subroutine

##### Arguments

`map`: shall be a scalar variable of type `open_hash_map_type`. It
is an `intent(inout)` argument. It will be a hash map used to store
and access the entry's data.

`inmap`: shall be a scalar integer expression of  kind `int_index`. It
is an `intent(in)` argument. It is the non-zero index in the inverse
table to the entry of interest.

* `inmap` will be invalid if zero, or `key` has been deleted from the
  map subsequent to the generation of `inmap`, or `map` has been
  rehashed subsequent to the generation of `inmap`.

`other`: shall be a scalar expression of type `other_type`.
It is an `intent(in)` argument. It is the data to be stored as
the other data for the entry at the `inmap` index.

* If  unable to set the other data associated with `inmap`, either
  because `inmap` is not associated with a valid entry or because of
  allocation problems, then processing will stop with an informative
  stop code.

##### Example

>     program demo_set_other_data

Author: I did a search for program  and added the Fortran code marking where missing.

their slot index for a hash map

##### Syntax

`Result = [[stdlib_hashmap_open:total_depth]]( map )`

##### Class

Pure function

##### Argument

`map`: shall be a scalar expression of type
`open_hash_map_type`. It is an `intent(in)` argument. It is the
hash map of interest.

##### Result character

The result is a scalar integer of kind `int_depth`.

##### Result value

The result is the total number of one's based offsets of slot entries
from their slot index the map.

##### Example

```fortran
    program demo_probes
      use stdlib_hashmap_open, only: &
         open_hash_map_type, init, int_index, &
         fnv_1_hasher, total_depth
      implicit none
      type(open_hash_map_type) :: map
      integer(int_depth) :: initial_depth
      call init( map, fnv_1_hasher )
      initial_depth = total_depth (map)
      print *, "Initial total depth =  ", initial_depth
    end program demo_probes
```

#### `unmap` - returns a copy of the key

##### Status

Experimental

##### Description

Returns a copy of the key associated with an index to the 
inverse table.

##### Syntax

`call [[stdlib_hashmap_open:unmap]]( map, inmap, key )`

##### Class

Subroutine

##### Arguments

> `map`: shall be a scalar expression of type `open_hash_map_type.

Author: fixed the two instances.

It is an `intent(in)` argument. It is the hash map whose entry
is unmapped.

`inmap`: shall be a scalar integer expression of kind `int_index`. It
is an `intent(in)` argument. It is the non-zero index to the inverse
table identifying the unmapped entry.

> * If ``inmap` is zero or `key` hass been eliminated from the table

Author: Done

  subsequent to the generation of `inmap`, or `map` has been rehashed
  subsequent to the generation of `inmap`, `other` is undefined.

`key`: shall be a variable of type `key_type`

> `INT8`, or an allocatable length default character. It is an

Author: Changed all instances of `INT8`, `INT32`, and `INT64` to `int8`, `int32`, and `int64` respectively.
